File: 0498-diagonal-traverse/0498-diagonal-traverse.js
/**
 * @param {number[][]} mat
 * @return {number[]}
 */
var findDiagonalOrder = function(mat) {
   let m = mat.length,
       n = mat[0].length;
    
    let arr = Array.from({ length: m + n - 1}, () => new Array());
    let result = new Array(m * n);
    let index = 0;
    
    for (let row = 0; row < m; ++row) {
        for (let col = 0; col < n; ++col) {
            arr[row + col].push(mat[row][col]);
        }
    }
    
    for (let d = 0; d < m + n - 1; ++d) {
        if (d % 2 == 0) {
            for (let i = arr[d].length - 1; i >= 0; --i) {
                result[index++] = arr[d][i];
            } 
        } else {
            for (let i = 0; i < arr[d].length; ++i) {
                result[index++] = arr[d][i];
            } 
        }
    }
    
    return result;
};

File: 0498-diagonal-traverse/0498-diagonal-traverse.java
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length,
            n = mat[0].length;
        List<Integer>[] dict = new ArrayList[m + n - 1];
        int[] result = new int[m * n];
        int index = 0;
        
        for (int i = 0; i < m + n - 1; ++i) {
            dict[i] = new ArrayList<Integer>();
        }
        
        for (int row = 0; row < m; ++row) {
            for (int col = 0; col < n; ++col) {
                dict[row + col].add(mat[row][col]);
            }
        }
     
        for (int i = 0; i < m + n - 1; ++i) {
            if (i % 2 == 0) {
                for (int j = dict[i].size() - 1; j >= 0; --j) {
                    result[index++] = dict[i].get(j);
                }
            } else {
                for (int j = 0; j < dict[i].size(); ++j) {
                    result[index++] = dict[i].get(j);
                }
            }
        }
        
        return result;
    }
}